core/src/main/java/io/questdb/KqueueFileWatcher.java

```diff
@@ -0,0 +1,219 @@
+            );
+
+            kq = KqueueAccessor.kqueue();
+            if (kq < 0) {
+                throw CairoException.critical(Os.errno()).put("could create kqueue");
+            }
+            Files.bumpFileCount(this.kq);
+
+            this.bufferSize = KqueueAccessor.SIZEOF_KEVENT;
+            this.eventList = Unsafe.calloc(bufferSize, MemoryTag.NATIVE_IO_DISPATCHER_RSS);
+
+            // Register events with queue
+            int fileRes = KqueueAccessor.keventRegister(
+                    kq,
+                    evtFile,
+                    1
+            );
+            if (fileRes < 0) {
+                throw CairoException.critical(Os.errno()).put("could register file events in kqueue");
+            }
+
+            int dirRes = KqueueAccessor.keventRegister(
+                    kq,
+                    evtDir,
+                    1
+            );
+            if (dirRes < 0) {
+                throw CairoException.critical(Os.errno()).put("could register directory events in kqueue");
+            }
+
+            int pipeRes = KqueueAccessor.keventRegister(
+                    kq,
+                    evtPipe,
+                    1
+            );
+            if (pipeRes < 0) {
+                throw CairoException.critical(Os.errno()).put("could register pipe events in kqueue");
+            }
+
+        } catch (RuntimeException e) {
+            cleanUp();
+            throw e;
+        }
+    }
+
+    private void cleanUp() {
+        if (kq > 0) {
+            Files.close(kq);
+        }
+
+        if (fileFd > 0) {
+            Files.close(fileFd);
+        }
+
+        if (dirFd > 0) {
+            Files.close(dirFd);
+        }
+
+        if (readEndFd > 0) {
+            Files.close(readEndFd);
+        }
+
+        if (writeEndFd > 0) {
+            Files.close(writeEndFd);
+        }
+
+        if (this.eventList > 0) {
+            Unsafe.free(this.eventList, bufferSize, MemoryTag.NATIVE_IO_DISPATCHER_RSS);
+        }
+
+        if (this.evtFile > 0) {
+            KqueueAccessor.evtFree(this.evtFile);
+        }
+
+        if (this.evtDir > 0) {
+            KqueueAccessor.evtFree(this.evtDir);
+        }
+
+        if (this.evtPipe > 0) {
+            KqueueAccessor.evtFree(this.evtPipe);
+        }
+    }
+
+    @Override
+    protected void _close() {
+        cleanUp();
+    }
+
+    @Override
+    protected void releaseWait() {
+        KqueueAccessor.writePipe(writeEndFd);
+    }
+
+    @Override
+    protected void waitForChange() {
+        // Blocks until there is a change in the watched dir
+        int res = KqueueAccessor.keventGetBlocking(
+                kq,
+                eventList,
+                1
+        );
+        if (res < 0) {
+            throw CairoException.critical(Os.errno()).put("kevent error");
+        }
+        // For now, we don't filter events down to the file,
+        // we trigger on every change in the directory
+        // todo: implement kqueue file filter
```

Comment:
do we need the filter at all? perhaps this TODO can be just removed?

Reply:
nope.. not for the current use case